Products/DataCollector/zing/DatamapHandler.py

```diff
@@ -88,21 +94,36 @@ def process_datamaps(self, tx_success, zing_state):
         if not tx_success or not self.zing_connector_url:
             return
         try:
-            log.info("Sending {} datamaps to zing-connector.".format(len(zing_state.datamaps)))
+            log.debug("Sending {} datamaps to zing-connector.".format(len(zing_state.datamaps)))
+            facts = []
             for device, datamap in zing_state.datamaps:
-                # TODO should we batch them instead of sending them one by one?
-                self._send_datamap(device, datamap, zing_state.contexts)
+                dm_facts = facts_from_datamap(device, datamap, zing_state.contexts)
+                if dm_facts:
+                    facts.extend(dm_facts)
+            # add dummy facts for the processed devices
+            device_facts = zing_state.devices_fact.values()
+            if device_facts:
+                facts.extend(device_facts)
+            self._send_facts_in_batches(facts)
         except Exception as ex:
             log.warn("Exception sending datamaps to zing-connector. {}".format(ex))
 
-    def _send_datamap(self, device, datamap, context):
+    def _send_facts_in_batches(self, facts, batch_size=500):
+        log.info("Sending {} facts to zing-connector.".format(len(facts)))
+        while facts:
+            batch = facts[:batch_size]
+            del facts[:batch_size]
+            self._send_facts(batch)
+
+    def _send_facts(self, facts):
         try:
-            dm = serialize_datamap(device, datamap, context)
-            if not dm: # nothing to send
+            serialized = serialize_facts(facts)
+            if not serialized: # nothing to send
```

Comment:
Do we need to log this as a warning or error? I'm assuming that normally we would always get a serialized version, right?

Reply:
that should be `if not facts`